test/distributed/org/apache/cassandra/distributed/test/cql3/StatefulASTBase.java

```diff
@@ -496,11 +499,43 @@ protected ByteBuffer[][] executeQuery(IInstance instance, int fetchSize, Consist
                                  .findAny()
                                  .get();
                 ss.setHost(host);
-                ResultSet result = session.execute(ss);
+                // Only SELECT is known safe to retry, as Mutations might not be idempotent!
+                boolean allowRetries = stmt.kind() == Statement.Kind.SELECT;
+                ResultSet result = !allowRetries ? session.execute(ss)
+                                                 : executeWithRetries(ss, 3);
                 return getRowsAsByteBuffer(result);
             }
         }
 
+        private ResultSet executeWithRetries(SimpleStatement ss, int maxTries)
+        {
+            Backoff backoff = null;
+            Throwable lastError = null;
+            for (int i = 0; i < maxTries; i++)
+            {
+                try
+                {
+                    return session.execute(ss);
+                }
+                catch (Throwable t)
```

Comment:
Is there a more specific exception to catch?

Reply:
Worried about catching `Throwable`s that are `Errors`s?

Reply:
Not a huge deal as not too extra time is being added for actual errors. It is just a matter of trying to be as specific as possible on the catch if we are adding this for the sole purpose of dealing with some ephemeral read errors.

Not a blocker for the PR, I already gave my +1 on the ticket when posting this question.

Reply:
switched to `ReadFailureException` and enriched the error to show what peer failed and why

Reply:
thanks for this feedback!  Made the change, which then gives me access to the error map (instance -> reason); so i enhanced the error messaging to show the reason

```
Caused by: java.lang.AssertionError: Request failed after 3 attempts! failed from={/127.0.0.3=INDEX_BUILD_IN_PROGRESS}
```

@maedhroz why are we rebuilding the index?!!?!?!?1

Reply:
so turns out this is basically a test bug

we use SAIUtils to see if its safe to use the index, but this only checks `node1`.

`node2` and `node3` *learn* about each other's status via gossip

this puts us in a race where you run a query on node2/3 and they have yet to see the status updated on the peers

I pushed a fix so the utils checks every node